main.go

```diff
@@ -18,11 +20,196 @@ import (
 	"github.com/nats-io/nats.go"
 	"github.com/sirupsen/logrus"
 	"github.com/spf13/viper"
+	"go.opentelemetry.io/otel"
+	semconv "go.opentelemetry.io/otel/semconv/v1.7.0"
+	"go.opentelemetry.io/otel/trace"
+	"google.golang.org/protobuf/encoding/protojson"
+	"google.golang.org/protobuf/proto"
 	"google.golang.org/protobuf/types/known/timestamppb"
 )
 
 var log = logging.Log.WithFields(logrus.Fields{"service": "discoenv-users-service"})
 
+// NATSMessageHandler is the handler type that we support. Should map to the nats.Handler
+// function signature func(*nats.Msg).
+type NATSMessageHandler func(context.Context, *nats.Msg)
+
+type NATSTextMapCarrier struct {
+	nats.Header
+}
+
+func (n NATSTextMapCarrier) Keys() []string {
+	var keys []string
+	for key := range n.Header {
+		keys = append(keys, key)
+	}
+	return keys
+}
+
+func ProtoUnmarshal(data []byte, ptr interface{}) error {
+	if _, ok := ptr.(*interface{}); ok {
+		return nil
+	}
+
+	msg, ok := ptr.(proto.Message)
+	if !ok {
+		return errors.New("invalid protocol buffer message passed to ProtoUnmarshal")
+	}
+
+	return protojson.Unmarshal(data, msg)
+}
+
+// func ProtoMarshal(ptr interface{}) ([]byte, error) {
+// 	msg, ok := ptr.(proto.Message)
+// 	if !ok {
+// 		return nil, errors.New("invalid protocol buffer message passed to ProtoMarshal")
+// 	}
+// 	b, err := protojson.Marshal(msg)
+// 	if err != nil {
+// 		return nil, err
+// 	}
+// 	return b, nil
+// }
+
+type NATSOtelEncodedConnection struct {
+	nats.EncodedConn
+	OtelInstrumentationName string
+	OtelSystem              string
+	OtelProtocol            string
+	OtelProtocolVersion     string
```

Comment:
Low priority, but these three things should probably be hardcoded and/or pulled from NATS code and not passed into anything a user's doing. Basically, shouldn't be possible to mispell "NATS" or have a mismatched version number compared to what the underlying NATS code actually is.

Reply:
This should be fixed now.